feat(gsvm): implemented GSVM v2 and labeled v1 as legacy

islerfab opened the pull request:
@beyeler , these are the changes which we discussed in the last call.

* GSVM adapted to the version that's closer to the original model's interpretation
* Added a flag that can be set to use the old version during model creation
* Set the MIPs' defaults to "not allowing assignments of licences with zero base values" (efficient allocation & demand query)

Closes #55

beyeler commented on src/main/java/org/spectrumauctions/sats/opt/model/gsvm/GSVMStandardMIP.java:
```diff
@@ -40,11 +40,7 @@ public GSVMStandardMIP(List<GSVMBidder> population) {
-	}
-
-	public GSVMStandardMIP(GSVMWorld world, List<GSVMBidder> population, boolean allowAssigningLicensesWithZeroBasevalue) {
-		this.allowAssigningLicensesWithZeroBasevalue = allowAssigningLicensesWithZeroBasevalue;
+		this.allowAssigningLicensesWithZeroBasevalue = world.isLegacyGSVM();
```
I changed the GSVMStandardMIP such that it responds to the worlds legacy setting. In this way it is possible to get the old behaviour by just changeing the world to legacy and efficient allocation, demand query and value query will work on the same value model.
Furthermore I added some tests for the demand query in legacy and non legacy setting.

islerfab replied:
Great addition! Before, there was the possibility of an inconsistency if the user plays around with the parameters. Now, the single source of truth is the initially created world.

islerfab commented:
@beyeler , if you think this PR is ready, please go ahead and merge it (I think `feat-integrate-mechlib-interaction` is not protected, right?)

beyeler commented on src/main/java/org/spectrumauctions/sats/core/model/gsvm/GSVMBidder.java:
```diff
@@ -46,15 +46,32 @@
+        
         double factor = 0;
-        if (!bundle.getBundleEntries().isEmpty()) factor = 0.2 * (bundle.getBundleEntries().size() - 1);
+        if (synergyCount > 0) factor = 0.2 * (synergyCount - 1);
```
I think the old version can lead to some problems. This way it is clear that no negative values will be returned also for bundles who only contain items with 0 base value

islerfab replied:
I agree! 👍 Thanks for the catch.

islerfab commented:
Good job. I agree on only keeping the two versions. I'm wondering if we could have a "default" for the activity limit that is not a number, but defined by the world. Thus, a national bidder would, by default, have an activity limit that is equal to the number of items in the national circle, and the regional bidder would have an activity limit that would be 2/3 (rounded) of the number of items he has a positive base value for. That way, it would scale reasonably by default when new worlds are created. Of course, a user should then still be able to set arbitrary activity limits, this would only concern the defaults.
What do you think?

beyeler commented:
Concerning the activity limit: 
I don't think that the model scales well to other worlds than the default. For other wolrds you would probably need to balance items, base values, synergies and activity limits manually to achive a balanced domain. Therefore this has no prioritiy to me.

islerfab commented:
I think it's important to make this activity limit scale with the world - I do I agree that other worlds may make less sense than the standard world, but GSVM is a particular example where the world is simple enough to make experiments with other parameters (e.g. the sizes of the circles can be set in SATS, so the activity limit should adapt). I added the implementation for that.

I also fixed the tests for that, and here I'm suddenly worried about our Value Model. According to the current implementation, if you ask a regional bidder what his value is for the six good he's interested in, he'll only tell you "the four highest-valued plus the synergy of having those four". This may be the way it was done in the experiments, but interpreted in a SATS-way this seems very odd - it's not actually his value, but given the activity rules, he would choose to only give the value for the four items.
If you guys decided that this is the way to go, I'm ok with it, but I fear we're making our future lives more difficult with including activity rules already in the value query...

beyeler commented:
@islerfab Did you push your changes concerning making the activity limits relative?

I can understand your concerns about the value model. But we thought this is the best way to have a model that behaves consistent no matter how you use it. You may also imagine that a regional bidder is only interested in 4 out of 6 licences. Then this would be the natural value model.

islerfab commented:
@beyeler I did, even though I can't explain the commit message for it... It was in this commit: https://github.com/spectrumauctions/sats/pull/58/commits/75c12f7998cb4267ce509733e96e172b1f893206

I am ok with the value model change, I just wanted to raise the point. Maybe, if activity rules will at some point be added to the MechLib / SATS, we can come back to this.

beyeler commented:
Ok perfect. Thx for the change. Looks good to me. I merge this request so we can discuss the next PR that depends on this branch without limitations.
